### PR TITLE
Switch over to coercion-based inlining

### PR DESCRIPTION
**Pull request by lukemaurer**
The decision logic is intended to be unchanged from the ancien régime,
with the exception that unrolling state is preserved between the
simplification of a recursive function's body and any re-simplification
that occurs due to inlining.

### CODE REVIEW

**Comment by mshinwell** [comment]
@lukemaurer Relating to the comment in this diff about something being semantically dodgy: has that been resolved by your conversation today with @lpw25 ?  If so please update the comment.

**Comment by lukemaurer** [comment]
> @lukemaurer Relating to the comment in this diff about something being semantically dodgy: has that been resolved by your conversation today with @lpw25 ? If so please update the comment.

Hmm ... it didn't answer the particular question in the comment, as far as I remember. The answer may similarly have to do with completing the lattice, though.

File: middle_end/flambda/inlining/inlining_decision.ml [hunk 1]
```diff
@@ -423,23 +423,63 @@ let might_inline dacc ~apply ~function_decl ~simplify_expr ~return_arity
...
-      if Inlining_state.is_depth_exceeded (Apply.inlining_state apply)
+    let unrolling_depth =
+      Simplify_rec_info_expr.known_remaining_unrolling_depth
+        dacc function_decl_rec_info
+    in
+    match unrolling_depth with
+    | Some 0 ->
+      Unrolling_depth_exceeded
+    | Some _ ->
+      might_inline dacc ~apply ~function_decl ~simplify_expr ~return_arity
```
Comment (poechsel) [inline middle_end/flambda/inlining/inlining_decision.ml]: Could you leave a comment to explain that at this point if `unrolling_depth` is `Some _` it means that the user annotated this function with an unroll annotation earlier on, meaning that we are sure that this function should be unrolled?

Comment (lukemaurer) [reply middle_end/flambda/inlining/inlining_decision.ml]: Sure, done.

**Review by poechsel — APPROVED** [review]

**Pull request merged**